### PR TITLE
Add auto spill ut

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #7738

Problem Summary:

### What is changed and how it works?
Add ut for auto spill
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Interpreters/SortSpillContext.cpp [hunk 1]
```diff
@@ -66,4 +80,17 @@ void SortSpillContext::finishOneSpill()
...
+bool SortSpillContext::needFinalSpill()
+{
+    if (auto_spill_status != AutoSpillStatus::NO_NEED_AUTO_SPILL)
+        return true;
+    bool ret = false;
+    auto old_value = AutoSpillStatus::NO_NEED_AUTO_SPILL;
+    fiu_do_on(FailPoints::random_marked_for_auto_spill, {
+        if (auto_spill_status.compare_exchange_strong(old_value, AutoSpillStatus::NEED_AUTO_SPILL))
+            ret = true;
+    });
```
Comment (SeaRise) [inline dbms/src/Interpreters/SortSpillContext.cpp:93]:
```suggestion
    fiu_do_on(FailPoints::random_marked_for_auto_spill, {
        auto old_value = AutoSpillStatus::NO_NEED_AUTO_SPILL;
        if (auto_spill_status.compare_exchange_strong(old_value, AutoSpillStatus::NEED_AUTO_SPILL))
            ret = true;
    });
```

File: dbms/src/Flash/Coprocessor/DAGContext.h [hunk 2]
```diff
@@ -344,7 +344,16 @@ class DAGContext
...
-    void setAutoSpillMode() { in_auto_spill_mode = true; }
+    void setAutoSpillMode()
+    {
+        in_auto_spill_mode = true;
+        /// only happens in test
+        if unlikely (query_operator_spill_contexts == nullptr)
+        {
+            query_operator_spill_contexts
+                = std::make_shared<QueryOperatorSpillContexts>(MPPQueryId(0, 0, 0, 0, ""), 100);
+        }
```
[the remark this answers is not in the document]
Comment (windtalker) [reply dbms/src/Flash/Coprocessor/DAGContext.h:355]: Done

File: dbms/src/Interpreters/SortSpillContext.cpp [hunk 3]
```diff
@@ -41,6 +47,14 @@ bool SortSpillContext::updateRevocableMemory(Int64 new_value)
             /// in auto spill mode, don't set revocable_memory to 0 here, so in triggerSpill it will take
             /// the revocable_memory into account if current spill is on the way
             return true;
+        old_value = AutoSpillStatus::NO_NEED_AUTO_SPILL;
```
[the remark this answers is not in the document]
Comment (windtalker) [reply dbms/src/Interpreters/SortSpillContext.cpp:50]: Done

File: dbms/src/Interpreters/SortSpillContext.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Interpreters/SortSpillContext.cpp:93]: Done

File: dbms/src/TestUtils/ExecutorTestUtils.cpp [hunk 4]
```diff
@@ -330,12 +330,20 @@ DB::ColumnsWithTypeAndName ExecutorTest::executeStreams(
     return executeStreams(&dag_context);
 }
 
-ColumnsWithTypeAndName ExecutorTest::executeStreams(DAGContext * dag_context)
+ColumnsWithTypeAndName ExecutorTest::executeStreamsForAutoSpill(
```
[the remark this answers is not in the document]
Comment (windtalker) [reply dbms/src/TestUtils/ExecutorTestUtils.cpp:333]: OK, I rename it to `executeStreamWithMemoryTracker`

**Review by SeaRise — APPROVED** [review]
lgtm

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**